File: multiple-languages/go/alicloudroscdkvpc/IVpnGateway.go
package alicloudroscdkvpc

import (
	_jsii_ "github.com/aws/jsii-runtime-go/runtime"

	"github.com/alibabacloud-go/ros-cdk/alicloudroscdkcore"
	"github.com/alibabacloud-go/ros-cdk/alicloudroscdkvpc/internal"
)

// Represents a `VpnGateway`.
type IVpnGateway interface {
	alicloudroscdkcore.IResource
	// Attribute Arn: The Alibaba Cloud Resource Name (ARN).
	AttrArn() interface{}
	// Attribute DisasterRecoveryInternetIp: The second IP address assigned by the system to create an IPsec-VPN connection.This attribute is returned only when the VPN gateway supports the dual-tunnel mode.
	AttrDisasterRecoveryInternetIp() interface{}
	// Attribute DisasterRecoveryVSwitchId: The ID of the second vSwitch associated with the VPN gateway.This attribute is returned only when the VPN gateway supports the dual-tunnel mode.
	AttrDisasterRecoveryVSwitchId() interface{}
	// Attribute InternetIp: The public IP address of the VPN gateway.
	AttrInternetIp() interface{}
	// Attribute OrderId: The order ID.
	AttrOrderId() interface{}
	// Attribute Spec: The specification of the VPN gateway.
	AttrSpec() interface{}
	// Attribute SslMaxConnections: The maximum number of concurrent SSL-VPN connections.
	AttrSslMaxConnections() interface{}
	// Attribute SslVpnInternetIp: The IP address of the SSL-VPN connection.This attribute is returned only when the VPN gateway is a public VPN gateway and supports only the single-tunnel mode. In addition, the VPN gateway must have the SSL-VPN feature enabled.
	AttrSslVpnInternetIp() interface{}
	// Attribute VpcId: The ID of the virtual private cloud (VPC) to which the VPN gateway belongs.
	AttrVpcId() interface{}
	// Attribute VpnGatewayId: ID of the VPN gateway.
	AttrVpnGatewayId() interface{}
	// Attribute VpnType: The type of the VPN gateway.
	AttrVpnType() interface{}
	// Attribute VSwitchId: The ID of the vSwitch to which the VPN gateway belongs.
	AttrVSwitchId() interface{}
	Props() *VpnGatewayProps
}

// The jsii proxy for IVpnGateway
type jsiiProxy_IVpnGateway struct {
	internal.Type__alicloudroscdkcoreIResource
}

func (j *jsiiProxy_IVpnGateway) AttrArn() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrArn",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrDisasterRecoveryInternetIp() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrDisasterRecoveryInternetIp",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrDisasterRecoveryVSwitchId() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrDisasterRecoveryVSwitchId",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrInternetIp() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrInternetIp",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrOrderId() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrOrderId",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrSpec() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrSpec",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrSslMaxConnections() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrSslMaxConnections",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrSslVpnInternetIp() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrSslVpnInternetIp",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrVpcId() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrVpcId",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrVpnGatewayId() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrVpnGatewayId",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrVpnType() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrVpnType",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) AttrVSwitchId() interface{} {
	var returns interface{}
	_jsii_.Get(
		j,
		"attrVSwitchId",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_IVpnGateway) Props() *VpnGatewayProps {
	var returns *VpnGatewayProps
	_jsii_.Get(
		j,
		"props",
		&returns,
	)
	return returns
}

File: multiple-languages/dotnet/AlibabaCloud.SDK.ROS.CDK.Vpc/AlibabaCloud/SDK/ROS/CDK/Vpc/IVpnGateway.cs
using Amazon.JSII.Runtime.Deputy;

#pragma warning disable CS0672,CS0809,CS1591

namespace AlibabaCloud.SDK.ROS.CDK.Vpc
{
    /// <summary>Represents a `VpnGateway`.</summary>
    [JsiiInterface(nativeType: typeof(IVpnGateway), fullyQualifiedName: "@alicloud/ros-cdk-vpc.IVpnGateway")]
    public interface IVpnGateway : AlibabaCloud.SDK.ROS.CDK.Core.IResource
    {
        /// <summary>Attribute Arn: The Alibaba Cloud Resource Name (ARN).</summary>
        [JsiiProperty(name: "attrArn", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrArn
        {
            get;
        }

        /// <summary>Attribute DisasterRecoveryInternetIp: The second IP address assigned by the system to create an IPsec-VPN connection.This attribute is returned only when the VPN gateway supports the dual-tunnel mode.</summary>
        [JsiiProperty(name: "attrDisasterRecoveryInternetIp", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrDisasterRecoveryInternetIp
        {
            get;
        }

        /// <summary>Attribute DisasterRecoveryVSwitchId: The ID of the second vSwitch associated with the VPN gateway.This attribute is returned only when the VPN gateway supports the dual-tunnel mode.</summary>
        [JsiiProperty(name: "attrDisasterRecoveryVSwitchId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrDisasterRecoveryVSwitchId
        {
            get;
        }

        /// <summary>Attribute InternetIp: The public IP address of the VPN gateway.</summary>
        [JsiiProperty(name: "attrInternetIp", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrInternetIp
        {
            get;
        }

        /// <summary>Attribute OrderId: The order ID.</summary>
        [JsiiProperty(name: "attrOrderId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrOrderId
        {
            get;
        }

        /// <summary>Attribute Spec: The specification of the VPN gateway.</summary>
        [JsiiProperty(name: "attrSpec", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrSpec
        {
            get;
        }

        /// <summary>Attribute SslMaxConnections: The maximum number of concurrent SSL-VPN connections.</summary>
        [JsiiProperty(name: "attrSslMaxConnections", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrSslMaxConnections
        {
            get;
        }

        /// <summary>Attribute SslVpnInternetIp: The IP address of the SSL-VPN connection.This attribute is returned only when the VPN gateway is a public VPN gateway and supports only the single-tunnel mode. In addition, the VPN gateway must have the SSL-VPN feature enabled.</summary>
        [JsiiProperty(name: "attrSslVpnInternetIp", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrSslVpnInternetIp
        {
            get;
        }

        /// <summary>Attribute VpcId: The ID of the virtual private cloud (VPC) to which the VPN gateway belongs.</summary>
        [JsiiProperty(name: "attrVpcId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrVpcId
        {
            get;
        }

        /// <summary>Attribute VpnGatewayId: ID of the VPN gateway.</summary>
        [JsiiProperty(name: "attrVpnGatewayId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrVpnGatewayId
        {
            get;
        }

        /// <summary>Attribute VpnType: The type of the VPN gateway.</summary>
        [JsiiProperty(name: "attrVpnType", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrVpnType
        {
            get;
        }

        /// <summary>Attribute VSwitchId: The ID of the vSwitch to which the VPN gateway belongs.</summary>
        [JsiiProperty(name: "attrVSwitchId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
        object AttrVSwitchId
        {
            get;
        }

        [JsiiProperty(name: "props", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-vpc.VpnGatewayProps\"}")]
        AlibabaCloud.SDK.ROS.CDK.Vpc.IVpnGatewayProps Props
        {
            get;
        }

        /// <summary>Represents a `VpnGateway`.</summary>
        [JsiiTypeProxy(nativeType: typeof(IVpnGateway), fullyQualifiedName: "@alicloud/ros-cdk-vpc.IVpnGateway")]
        internal sealed class _Proxy : DeputyBase, AlibabaCloud.SDK.ROS.CDK.Vpc.IVpnGateway
        {
            private _Proxy(ByRefValue reference): base(reference)
            {
            }

            /// <summary>Attribute Arn: The Alibaba Cloud Resource Name (ARN).</summary>
            [JsiiProperty(name: "attrArn", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrArn
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute DisasterRecoveryInternetIp: The second IP address assigned by the system to create an IPsec-VPN connection.This attribute is returned only when the VPN gateway supports the dual-tunnel mode.</summary>
            [JsiiProperty(name: "attrDisasterRecoveryInternetIp", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrDisasterRecoveryInternetIp
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute DisasterRecoveryVSwitchId: The ID of the second vSwitch associated with the VPN gateway.This attribute is returned only when the VPN gateway supports the dual-tunnel mode.</summary>
            [JsiiProperty(name: "attrDisasterRecoveryVSwitchId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrDisasterRecoveryVSwitchId
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute InternetIp: The public IP address of the VPN gateway.</summary>
            [JsiiProperty(name: "attrInternetIp", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrInternetIp
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute OrderId: The order ID.</summary>
            [JsiiProperty(name: "attrOrderId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrOrderId
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute Spec: The specification of the VPN gateway.</summary>
            [JsiiProperty(name: "attrSpec", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrSpec
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute SslMaxConnections: The maximum number of concurrent SSL-VPN connections.</summary>
            [JsiiProperty(name: "attrSslMaxConnections", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrSslMaxConnections
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute SslVpnInternetIp: The IP address of the SSL-VPN connection.This attribute is returned only when the VPN gateway is a public VPN gateway and supports only the single-tunnel mode. In addition, the VPN gateway must have the SSL-VPN feature enabled.</summary>
            [JsiiProperty(name: "attrSslVpnInternetIp", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrSslVpnInternetIp
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute VpcId: The ID of the virtual private cloud (VPC) to which the VPN gateway belongs.</summary>
            [JsiiProperty(name: "attrVpcId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrVpcId
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute VpnGatewayId: ID of the VPN gateway.</summary>
            [JsiiProperty(name: "attrVpnGatewayId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrVpnGatewayId
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute VpnType: The type of the VPN gateway.</summary>
            [JsiiProperty(name: "attrVpnType", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrVpnType
            {
                get => GetInstanceProperty<object>()!;
            }

            /// <summary>Attribute VSwitchId: The ID of the vSwitch to which the VPN gateway belongs.</summary>
            [JsiiProperty(name: "attrVSwitchId", typeJson: "{\"union\":{\"types\":[{\"primitive\":\"string\"},{\"fqn\":\"@alicloud/ros-cdk-core.IResolvable\"}]}}")]
            public object AttrVSwitchId
            {
                get => GetInstanceProperty<object>()!;
            }

            [JsiiProperty(name: "props", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-vpc.VpnGatewayProps\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Vpc.IVpnGatewayProps Props
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Vpc.IVpnGatewayProps>()!;
            }

            /// <summary>The environment this resource belongs to.</summary>
            /// <remarks>
            /// For resources that are created and managed by the CDK
            /// (generally, those created by creating new class instances like Role, Bucket, etc.),
            /// this is always the same as the environment of the stack they belong to;
            /// however, for imported resources
            /// (those obtained from static methods like fromRoleArn, fromBucketName, etc.),
            /// that might be different than the stack they were imported into.
            /// </remarks>
            [JsiiProperty(name: "env", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.IResourceEnvironment\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Core.IResourceEnvironment Env
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.IResourceEnvironment>()!;
            }

            /// <summary>The stack in which this resource is defined.</summary>
            [JsiiProperty(name: "stack", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.Stack\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Core.Stack Stack
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.Stack>()!;
            }

            /// <summary>The construct tree node for this construct.</summary>
            [JsiiProperty(name: "node", typeJson: "{\"fqn\":\"@alicloud/ros-cdk-core.ConstructNode\"}")]
            public AlibabaCloud.SDK.ROS.CDK.Core.ConstructNode Node
            {
                get => GetInstanceProperty<AlibabaCloud.SDK.ROS.CDK.Core.ConstructNode>()!;
            }
        }
    }
}
